### PR TITLE
Make NthValue templated for int or bigint offset use

### PR DESCRIPTION
**Pull request by pramodsatya**
NthValue had some type specific code for int and bigint offset parameter. The int and bigint signatures were also not consistent as only constant offsets were supported for integer whereas both constant and column offsets were supported for bigint offset. So there were some engine specific error messages added in the code.

This PR makes the implementation between int and bigint uniform. Velox has the superset of function usage for Presto/Spark.

### CODE REVIEW

**Comment by aditi-pandit** [comment]
@pramodsatya : This is not the functionality I expected. The goal was to allow both int32 and int64 constants or columns for offset parameter in this function. Spark would then use the correct type as required.

**Comment by pramodsatya** [comment]
> @pramodsatya : This is not the functionality I expected. The goal was to allow both int32 and int64 constants or columns for offset parameter in this function. Spark would then use the correct type as required.

Thanks for sharing @aditi-pandit, I will update the PR accordingly. Is there an existing check in Spark that deters the use of non int32 constant as offset?

**Comment by aditi-pandit** [comment]
> Reference

@pramodsatya : Yes, I believe Spark invalidates for a int64 column offset.

File: velox/functions/sparksql/window/tests/NthValueTest.cpp [hunk 1]
```diff
@@ -64,34 +64,23 @@ TEST_F(SparkNthValueTest, singlePartition) {
       {vectors}, "nth_value(c0, 1)", {overClause}, {frameClausee});
 }
```
Comment (aditi-pandit) [inline velox/functions/sparksql/window/tests/NthValueTest.cpp:65]: Please add a test in this file where int32 column is used for an offset parameter. Spark should not end up using it. But we should have a tested code-path.

**Review by aditi-pandit — COMMENTED** [review]
Thanks @pramodsatya.

Have a comment about the test. The rest of the code looks alright.

File: velox/functions/sparksql/window/tests/NthValueTest.cpp (hunk 1 above)
Comment (pramodsatya) [reply velox/functions/sparksql/window/tests/NthValueTest.cpp]: These changes were reverted recently and the tests are now moved SparkWindowTest.cpp.

Comment (aditi-pandit) [reply velox/functions/sparksql/window/tests/NthValueTest.cpp]: The intent was to ensure that velox function with int32 column offset works. This will not be used by Spark eval, but we should have a Velox test for it. Right now there aren't tests that use int32 column imo.

Velox unit test shouldn't be throwing any error. Spark will do its own error handling. And we don't use a Spark env in test framework.

Comment (aditi-pandit) [reply velox/functions/sparksql/window/tests/NthValueTest.cpp]: Are there any tests in SparkWindowTest.cpp for exercising the codepaths for int32 column offsets ?

Comment (pramodsatya) [reply velox/functions/sparksql/window/tests/NthValueTest.cpp]: Thanks for the clarification. There were no tests to check for int32 column offset in SparkWindowTest.cpp and it is added now.

File: velox/functions/lib/window/NthValue.cpp [hunk 2]
```diff
@@ -292,13 +276,14 @@ class NthValueFunction : public exec::WindowFunction {
...
-void registerNthValue(const std::string& name, TypeKind offsetTypeKind) {
+template <typename T>
+void registerNthValue(const std::string& name) {
   std::vector<exec::FunctionSignaturePtr> signatures{
       exec::FunctionSignatureBuilder()
           .typeVariable("T")
           .returnType("T")
           .argumentType("T")
-          .argumentType(mapTypeKindToName(offsetTypeKind))
+          .argumentType(CppToType<T>::name)
```
[the remark this answers is not in the document]
Comment (pramodsatya) [reply velox/functions/lib/window/NthValue.cpp]: Thanks for pointing it out, using logical types like DATE for offset currently results in an error from `createWindowFunctions::exprToChannel`. Reverted this change to ensure function registration happens for integer and bigint offsets.

**Review by aditi-pandit — APPROVED** [review]
Thanks @pramodsatya. This looks good.

**Comment by aditi-pandit** [comment]
@mbasmanova : This helps cleanup some code in NthValue that had resulted in some Spark specific behavior. The current Velox function is now a super-set of what is expected in Presto and Spark which is more inline with the design philosophy.
PTAL.

**Review by aditi-pandit — CHANGES_REQUESTED** [review]
@pramodsatya : Does the function need to be templatized globally ? Can we maintain separate int32/int64 constantOffset and offsets_ vectors and templatize only a subset of the functions like setRowNumber and setRowNumberIgnoreNull, setRowNumbersApplyLoop to achieve the same effect ?

File: velox/functions/lib/window/NthValue.cpp [hunk 3]
```diff
@@ -28,39 +28,46 @@ class NthValueFunction : public exec::WindowFunction {
...
+        VELOX_FAIL("Invalid offset type: {}", offsetType->toString());
+      }
+    } else {
+      offsetIndex_ = args[1].index.value();
+      if (offsetType->isInteger()) {
+        offsets_ = BaseVector::create<FlatVector<int32_t>>(offsetType, 0, pool);
+      } else if (offsetType->isBigint()) {
+        offsets_ = BaseVector::create<FlatVector<int64_t>>(offsetType, 0, pool);
+      } else {
+        VELOX_FAIL("Invalid offset type: {}", offsetType->toString());
```
Comment (karteekmurthys) [inline velox/functions/lib/window/NthValue.cpp:70]: L61 and L70 checks can be moved to the begining of the function VELOX_CHECK(type is int || type bigint)

File: velox/functions/lib/window/NthValue.cpp [hunk 4]
```diff
@@ -265,13 +299,14 @@ class NthValueFunction : public exec::WindowFunction {
 
   const exec::WindowPartition* partition_;
 
-  // These fields are set if the offset argument is a constant value.
-  std::optional<int64_t> constantOffset_;
+  // This field is set if the offset argument is a constant value.
+  std::optional<std::variant<int32_t, int64_t>> constantOffset_;
```
Comment (karteekmurthys) [inline velox/functions/lib/window/NthValue.cpp:303]: What breaks if we save int32 offset in a  `int64_t constantOffset_`.
We should be able to save int32 int it, right?

Isn't just adding signature sufficient, then?

File: velox/functions/lib/window/NthValue.cpp [hunk 5]
```diff
@@ -265,13 +299,14 @@ class NthValueFunction : public exec::WindowFunction {
...
   bool isConstantOffsetNull_ = false;
 
   // This vector is used to extract values of the offset argument column
   // (if not a constant offset value).
-  FlatVectorPtr<int64_t> offsets_;
+  std::optional<std::variant<FlatVectorPtr<int32_t>, FlatVectorPtr<int64_t>>>
```
Comment (karteekmurthys) [inline velox/functions/lib/window/NthValue.cpp:308]: Can you try BaseVector* and `dynamic_cast<>` at runtime.
Or, just like the above comment have a single FlatVector<int64_t>, will that cause any issues?

File: velox/functions/lib/window/NthValue.cpp (hunk 4 above)
Comment (pramodsatya) [reply velox/functions/lib/window/NthValue.cpp:303]: Thanks for the feedback, nothing would break as such because the function signature ensures the offset is of the right type in Spark/Presto. It would be better to represent an int32_t offset with the right type, instead of casting from an int64_t variable. 
What do you suggest @aditi-pandit ?

File: velox/functions/lib/window/tests/WindowTestBase.cpp [hunk 6]
```diff
@@ -100,7 +100,7 @@ RowVectorPtr WindowTestBase::makeRandomInputVector(vector_size_t size) {
       {makeRandomInputVector(BIGINT(), size, 0.2),
        makeRandomInputVector(VARCHAR(), size, 0.3),
        makeFlatVector<int64_t>(size, genRandomFrameValue),
-       makeFlatVector<int64_t>(size, genRandomFrameValue)});
+       makeFlatVector<int32_t>(size, genRandomFrameValue)});
```
Comment (mbasmanova) [inline velox/functions/lib/window/tests/WindowTestBase.cpp:103]: Is this a fix for unrelated issue? Can it be extracted into a separate PR?

File: velox/functions/lib/window/NthValue.cpp [hunk 7]
```diff
@@ -28,39 +28,46 @@ class NthValueFunction : public exec::WindowFunction {
...
-              ->valueAt(0);
-      VELOX_USER_CHECK_GE(
-          constantOffset_.value(), 1, "Offset must be at least 1");
-    } else {
-      if (args[1].constantValue) {
-        if (args[1].constantValue->isNullAt(0)) {
-          isConstantOffsetNull_ = true;
-          return;
-        }
+      if (offsetType->isInteger()) {
```
Comment (mbasmanova) [inline velox/functions/lib/window/NthValue.cpp:42]: perhaps, extract this logic into a helper function for readability

**Review by mbasmanova — COMMENTED** [review]
@pramodsatya Would you clarify the motivation for this change?

**Comment by pramodsatya** [comment]
@mbasmanova, this change makes the nth_value implementation for bigint and integer offsets, for constants and columns, uniform. The spark specific error message for constant integer offset is removed in the process as this check should not be in Velox.

File: velox/functions/lib/window/tests/WindowTestBase.cpp (hunk 6 above)
Comment (pramodsatya) [reply velox/functions/lib/window/tests/WindowTestBase.cpp:103]: This change was made to test the columnar int32_t offset in spark tests (`nth_value(c0, c3)`), since the check for constant integer type offset in spark is removed in this PR. Please let me know if it should be reverted.

File: velox/functions/lib/window/NthValue.cpp [hunk 8]
```diff
@@ -101,6 +90,29 @@ class NthValueFunction : public exec::WindowFunction {
   }
 
  private:
+  // Helper function to update the constant offset value.
+  void updateConstantOffset(
+      const exec::WindowFunctionArg& offsetArg,
+      const TypePtr& offsetType) {
+    if (offsetType->isInteger()) {
+      constantOffset_ =
```
Comment (karteekmurthys) [inline velox/functions/lib/window/NthValue.cpp:98]: At least for constantOffset_ we must stick just int64_t. Keeps it simple.
May be for vector offsets having 2 different sizes is a space optimization.

Comment (mbasmanova) [reply velox/functions/lib/window/NthValue.cpp:98]: @karteekmurthys Why do we have to use 64-bit integer for constant offset. Is it actually possible to have an offset that > max 32-bit integer?

File: velox/functions/lib/window/NthValue.cpp [hunk 9]
```diff
@@ -101,6 +90,29 @@ class NthValueFunction : public exec::WindowFunction {
...
+              ->valueAt(0);
+      VELOX_USER_CHECK_GE(
+          std::get<int32_t>(constantOffset_.value()),
+          1,
+          "Offset must be at least 1");
+    } else {
+      constantOffset_ =
+          offsetArg.constantValue->template as<ConstantVector<int64_t>>()
+              ->valueAt(0);
+      VELOX_USER_CHECK_GE(
```
Comment (mbasmanova) [inline velox/functions/lib/window/NthValue.cpp:109]: This check can be moved outside of the if-else block. No need to repeat it.

File: velox/functions/lib/window/NthValue.cpp (hunk 4 above)
Comment (mbasmanova) [reply velox/functions/lib/window/NthValue.cpp:308]: I suggest to use int64_t and not use std::variant.

File: velox/functions/lib/window/NthValue.cpp (hunk 8 above)
Comment (karteekmurthys) [reply velox/functions/lib/window/NthValue.cpp:98]: The reason 32-bit offset was added was to support Spark NthValue. I suggested to remove to reduce `if-else` checks at runtime. The Signature for PRESTO requires BIGINT while SPARK only INTEGER().

Comment (mbasmanova) [reply velox/functions/lib/window/NthValue.cpp:98]: Yes, I understand that Presto use BIGINT to represent offset, but can this value actually be bigger than max 32-bit integer. I assume not.

File: velox/functions/lib/window/NthValue.cpp [hunk 10]
```diff
@@ -305,7 +345,7 @@ void registerNthValue(const std::string& name, TypeKind offsetTypeKind) {
   exec::registerWindowFunction(
       name,
       std::move(signatures),
-      [name](
+      [name, offsetTypeKind](
```
Comment (aditi-pandit) [inline velox/functions/lib/window/NthValue.cpp:348]: Does offsetTypeKind need to be a parameter ? You can obtain it in the constructor using the type of the second argument.

File: velox/functions/lib/window/NthValue.cpp (hunk 4 above)
Comment (aditi-pandit) [reply velox/functions/lib/window/NthValue.cpp:308]: +1. We could just use int64 for the rest of the code.

File: velox/functions/lib/window/NthValue.cpp [hunk 11]
```diff
@@ -265,13 +304,14 @@ class NthValueFunction : public exec::WindowFunction {
...
-  // These fields are set if the offset argument is a constant value.
-  std::optional<int64_t> constantOffset_;
+  // This field is set if the offset argument is a constant value.
+  std::optional<std::variant<int32_t, int64_t>> constantOffset_;
   bool isConstantOffsetNull_ = false;
 
   // This vector is used to extract values of the offset argument column
   // (if not a constant offset value).
-  FlatVectorPtr<int64_t> offsets_;
+  std::optional<std::variant<FlatVectorPtr<int32_t>, FlatVectorPtr<int64_t>>>
```
Comment (aditi-pandit) [inline velox/functions/lib/window/NthValue.cpp:313]: Lets not use variant here either. We can have a BaseVector* here and extract the type from it at runtime.

File: velox/functions/lib/window/NthValue.cpp [hunk 12]
```diff
@@ -126,35 +138,54 @@ class NthValueFunction : public exec::WindowFunction {
...
+            numRows,
+            validRows,
+            rawFrameStarts,
+            rawFrameEnds,
+            leastFrame);
+      }
     }
   }
 
+  template <typename T>
```
Comment (aditi-pandit) [inline velox/functions/lib/window/NthValue.cpp:177]: This needn't be parameterized if constant offset is always int64.

File: velox/functions/lib/window/NthValue.cpp [hunk 13]
```diff
@@ -126,35 +138,54 @@ class NthValueFunction : public exec::WindowFunction {
...
+      }
     } else {
-      setRowNumbersForColumnOffset(
-          ignoreNullsForBlock,
-          numRows,
-          validRows,
-          rawFrameStarts,
-          rawFrameEnds,
-          leastFrame);
+      if (std::holds_alternative<FlatVectorPtr<int32_t>>(offsets_.value())) {
```
Comment (aditi-pandit) [inline velox/functions/lib/window/NthValue.cpp:157]: Do we need to use std::holds_alternative ? We can get the type of the vector from it directly.

File: velox/functions/lib/window/NthValue.cpp (hunk 8 above)
Comment (aditi-pandit) [reply velox/functions/lib/window/NthValue.cpp:98]: @mbasmanova : Do we want to make that assumption ? Our code uses vector_size_t overall for partition sizes, but we could have massive partitions also. wdyt ?

File: velox/functions/lib/window/NthValue.cpp (hunk 12 above)
Comment (pramodsatya) [reply velox/functions/lib/window/NthValue.cpp]: Reverted, thanks

File: velox/functions/lib/window/NthValue.cpp (hunk 4 above)
Comment (pramodsatya) [reply velox/functions/lib/window/NthValue.cpp]: Updated.

File: velox/functions/lib/window/NthValue.cpp (hunk 10 above)
Comment (pramodsatya) [reply velox/functions/lib/window/NthValue.cpp]: Removed, thanks

File: velox/functions/lib/window/NthValue.cpp (hunk 13 above)
Comment (pramodsatya) [reply velox/functions/lib/window/NthValue.cpp]: Reverted use of variants

**Review by karteekmurthys — APPROVED** [review]
Thanks for addressing the comments. LGTM!

**Comment by yzhang1991** [comment]
@aditi-pandit @pramodsatya Whats our plan with this PR?

**Review by aditi-pandit — APPROVED** [review]
Thanks @pramodsatya for the changes.

**Comment by aditi-pandit** [comment]
@mbasmanova : PTAL.

File: velox/functions/lib/window/NthValue.cpp [hunk 14]
```diff
@@ -32,34 +32,35 @@ class NthValueFunction : public exec::WindowFunction {
       : WindowFunction(resultType, pool, nullptr), ignoreNulls_(ignoreNulls) {
     VELOX_CHECK_EQ(args.size(), 2);
     VELOX_CHECK_NULL(args[0].constantValue);
+    auto offsetType = args[1].type;
+    VELOX_CHECK(
```
Comment (mbasmanova) [inline velox/functions/lib/window/NthValue.cpp:36]: USER_CHECK

**Review by mbasmanova — APPROVED** [review]
Thanks.

File: velox/functions/lib/window/NthValue.cpp (hunk 14 above)
Comment (pramodsatya) [reply velox/functions/lib/window/NthValue.cpp:36]: Thanks for the suggestion, but I don't have access to this upstream branch anymore. Would it be fine if this change is made in a subsequent PR?

Comment (mbasmanova) [reply velox/functions/lib/window/NthValue.cpp:36]: @pramodsatya Can you maybe create a new PR and close this one? Use gh pr checkout 6186.

Comment (pramodsatya) [reply velox/functions/lib/window/NthValue.cpp:36]: Sure, opened #8054 and closing this, thanks.
